markets/perps-market/contracts/storage/PerpsMarket.sol

```diff
@@ -96,77 +99,121 @@ library PerpsMarket {
-                liquidatableAmount = (
-                    requestedLiquidationAmount > maxAllowedLiquidationInWindow
-                        ? maxAllowedLiquidationInWindow
-                        : requestedLiquidationAmount
-                ).to128();
-                // track how much was utilized in this block to ensure any subsequent
-                // liquidations don't combine for larger than what's allotted in the window
-                self.lastUtilizedLiquidationCapacity = liquidatableAmount;
+                liquidatableAmount = MathUtil.min128(
+                    maxLiquidationInWindow.to128(),
+                    requestedLiquidationAmount
+                );
             }
         }
 
-        // only update timestamp if there is something being liquidated
         if (liquidatableAmount > 0) {
-            self.lastTimeLiquidationCapacityUpdated = block.timestamp.to128();
+            // if other liquidations happened in same block, add to amount otherwise set as liquidatable amt
+            self.liquidationAmounts[liquidationIndex] = self.liquidationTimestamps[
+                liquidationIndex
+            ] == block.timestamp
+                ? self.liquidationAmounts[liquidationIndex] + liquidatableAmount
+                : liquidatableAmount;
+            self.liquidationTimestamps[liquidationIndex] = block.timestamp.to128();
         }
     }
 
-    struct MarketUpdateData {
-        uint128 marketId;
-        int256 skew;
-        uint256 size;
-        int256 currentFundingRate;
-        int256 currentFundingVelocity;
+    /**
+     * @dev Returns the current liquidation capacity for the market
+     * @notice liquidationAmounts tracks every liquidation with the corresponding timestamp.
+     * The logic here goes through each item in the array and sums up the amounts that are within the
+     * configured liquidation window from the current block timestamp.  This value is used to determine
+     * the available capacity
+     */
+    function currentLiquidationCapacity(
+        Data storage self,
+        PerpsMarketConfiguration.Data storage marketConfig
+    )
+        internal
+        view
+        returns (uint capacity, uint256 maxLiquidationInWindow, uint128 latestLiquidationTimestamp)
+    {
+        uint accumulatedLiquidationAmounts;
+
+        uint maxConfiguredSeconds = marketConfig.maxSecondsInLiquidationWindow;
+        maxLiquidationInWindow = marketConfig.maxLiquidationAmountInWindow();
+
+        for (uint i = 0; i < maxConfiguredSeconds; i++) {
+            uint128 liquidationTimestamp = self.liquidationTimestamps[i];
+            uint windowStartTimestamp = block.timestamp.to128() -
+                marketConfig.maxSecondsInLiquidationWindow;
+
+            // keep track of latest liquidation timestamp
+            // used in determining whether to allow another block of liquidation given p/d requirements are met
+            latestLiquidationTimestamp = MathUtil
```

Comment:
do we need to do that in the loop? cannot we do the same .max with current timestamp and latestLiquidationTimestamp if capacity is > 0 at the end of this function?

Reply:
yeah we could.  the reason i have it here is because, if there is no liquidation capacity, this returns 0, and since there's a view called `liquidationCapacity(marketId)`, it returns 0 as the latestLIquidationTimestamp isntead of whatever the latest is regardless of if there's capacity.